File: arkanoid.js
// arkanoid.js
const readline = require('readline');
const { stdin, stdout } = process;

const W = 40, H = 20;
let paddleX = Math.floor(W/2) - 4;
const paddleWidth = 8;
let ballX = paddleX + Math.floor(paddleWidth/2);
let ballY = H - 3;
let ballDX = 1;
let ballDY = -1;
let ballLaunched = false;
let lives = 3;
let score = 0;
let level = 1;
let bricks = [];
let gameOver = false;
let running = true;

function initBricks() {
    bricks = [];
    const rows = Math.min(3 + level, 6);
    const cols = Math.floor(W / 3);
    for (let y = 0; y < rows; y++) {
        for (let x = 1; x < cols; x++) {
            bricks.push({ x: x * 3, y: y + 2, alive: true });
        }
    }
}

function drawBorder() {
    let out = '';
    for (let x = 0; x < W; x++) out += '#';
    out += '\n';
    for (let y = 1; y < H-1; y++) {
        out += '#';
        for (let x = 1; x < W-1; x++) out += ' ';
        out += '#\n';
    }
    for (let x = 0; x < W; x++) out += '#';
    return out;
}

function render() {
    console.clear();
    // Build grid
    let grid = Array.from({length: H}, () => Array(W).fill(' '));
    // border
    for (let x = 0; x < W; x++) { grid[0][x] = '#'; grid[H-1][x] = '#'; }
    for (let y = 0; y < H; y++) { grid[y][0] = '#'; grid[y][W-1] = '#'; }
    // paddle
    for (let x = 0; x < paddleWidth; x++) {
        grid[H-2][paddleX + x] = '=';
    }
    // ball
    if (ballLaunched || !gameOver) {
        grid[ballY][ballX] = 'O';
    }
    // bricks
    bricks.forEach(b => { if (b.alive) grid[b.y][b.x] = '█'; });
    // print
    console.log(grid.map(row => row.join('')).join('\n'));
    console.log(`Score: ${score}   Lives: ${lives}   Level: ${level}`);
    if (gameOver) console.log('GAME OVER! Press R');
    else if (!ballLaunched) console.log('Press Space to launch');
}

function updateBall() {
    if (!ballLaunched || gameOver) return;
    ballX += ballDX;
    ballY += ballDY;
    // walls
    if (ballX <= 1 || ballX >= W-2) ballDX *= -1;
    if (ballY <= 1) ballDY *= -1;
    // paddle
    if (ballY === H-2 && ballX >= paddleX && ballX < paddleX + paddleWidth) {
        ballDY *= -1;
        const hitPos = ballX - (paddleX + Math.floor(paddleWidth/2));
        if (hitPos !== 0) {
            ballDX = hitPos > 0 ? 1 : -1;
        }
    }
    // bricks
    for (let i = 0; i < bricks.length; i++) {
        if (bricks[i].alive && ballY === bricks[i].y && ballX === bricks[i].x) {
            bricks[i].alive = false;
            score += 10;
            ballDY *= -1;
            break;
        }
    }
    // bottom
    if (ballY >= H-1) {
        lives--;
        ballLaunched = false;
        ballX = paddleX + Math.floor(paddleWidth/2);
        ballY = H - 3;
        ballDX = Math.random() > 0.5 ? 1 : -1;
        ballDY = -1;
        if (lives === 0) gameOver = true;
    }
    // level complete
    if (bricks.every(b => !b.alive)) {
        level++;
        initBricks();
        ballLaunched = false;
        ballX = paddleX + Math.floor(paddleWidth/2);
        ballY = H - 3;
        ballDX = Math.random() > 0.5 ? 1 : -1;
        ballDY = -1;
    }
}

function movePaddle(dx) {
    if (gameOver) return;
    const newX = paddleX + dx;
    if (newX >= 1 && newX + paddleWidth <= W-1) {
        paddleX = newX;
        if (!ballLaunched) {
            ballX = paddleX + Math.floor(paddleWidth/2);
        }
    }
}

function launchBall() {
    if (gameOver) return;
    if (!ballLaunched) ballLaunched = true;
}

function restart() {
    paddleX = Math.floor(W/2) - 4;
    ballX = paddleX + Math.floor(paddleWidth/2);
    ballY = H - 3;
    ballDX = 1;
    ballDY = -1;
    ballLaunched = false;
    lives = 3;
    score = 0;
    level = 1;
    gameOver = false;
    initBricks();
}

function setupInput() {
    readline.emitKeypressEvents(process.stdin);
    process.stdin.setRawMode(true);
    process.stdin.on('keypress', (str, key) => {
        if (key.ctrl && key.name === 'c') process.exit();
        if (key.name === 'q') process.exit();
        if (key.name === 'r' && gameOver) restart();
        if (key.name === 'space') launchBall();
        if (key.name === 'left' || key.name === 'a') movePaddle(-2);
        if (key.name === 'right' || key.name === 'd') movePaddle(2);
    });
}

setupInput();
initBricks();
setInterval(() => {
    updateBall();
    render();
}, 50);

File: arkanoid.py
# arkanoid.py
import curses
import random

class Arkanoid:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        curses.curs_set(0)
        self.stdscr.nodelay(1)
        self.stdscr.timeout(50)
        self.height, self.width = self.stdscr.getmaxyx()
        self.width = min(self.width, 80)
        self.height = min(self.height, 24)
        self.paddle_x = self.width // 2 - 4
        self.paddle_width = 8
        self.ball_x = self.paddle_x + self.paddle_width // 2
        self.ball_y = self.height - 3
        self.ball_dx = 1
        self.ball_dy = -1
        self.ball_launched = False
        self.lives = 3
        self.score = 0
        self.level = 1
        self.bricks = []
        self.init_bricks()
        self.game_over = False
        self.paused = False
        self.running = True

    def init_bricks(self):
        self.bricks = []
        rows = min(3 + self.level, 6)
        cols = self.width // 3
        for y in range(rows):
            for x in range(1, cols):
                self.bricks.append([x*3, y+2, 1])  # x, y, alive

    def draw_border(self):
        for x in range(self.width):
            self.stdscr.addch(0, x, '#')
        for y in range(self.height):
            self.stdscr.addch(y, 0, '#')
            self.stdscr.addch(y, self.width-1, '#')

    def draw_paddle(self):
        for x in range(self.paddle_width):
            self.stdscr.addch(self.height-2, self.paddle_x + x, '=')

    def draw_ball(self):
        self.stdscr.addch(self.ball_y, self.ball_x, 'O')

    def draw_bricks(self):
        for x, y, alive in self.bricks:
            if alive:
                self.stdscr.addch(y, x, '█')

    def draw_info(self):
        self.stdscr.addstr(0, self.width+2, f"Score: {self.score}")
        self.stdscr.addstr(1, self.width+2, f"Lives: {self.lives}")
        self.stdscr.addstr(2, self.width+2, f"Level: {self.level}")
        if self.game_over:
            self.stdscr.addstr(self.height//2, self.width//2-5, "GAME OVER! Press R")
        elif not self.ball_launched:
            self.stdscr.addstr(self.height//2, self.width//2-5, "Press Space to launch")

    def update_ball(self):
        if not self.ball_launched or self.game_over:
            return
        self.ball_x += self.ball_dx
        self.ball_y += self.ball_dy
        # Wall collisions
        if self.ball_x <= 1 or self.ball_x >= self.width-2:
            self.ball_dx *= -1
        if self.ball_y <= 1:
            self.ball_dy *= -1
        # Paddle collision
        if self.ball_y == self.height-2 and self.paddle_x <= self.ball_x < self.paddle_x + self.paddle_width:
            self.ball_dy *= -1
            # Angle based on paddle hit position
            hit_pos = self.ball_x - (self.paddle_x + self.paddle_width//2)
            if hit_pos != 0:
                self.ball_dx = 1 if hit_pos > 0 else -1
        # Brick collision
        for i, (x, y, alive) in enumerate(self.bricks):
            if alive and self.ball_y == y and self.ball_x == x:
                self.bricks[i][2] = 0
                self.score += 10
                self.ball_dy *= -1
                break
        # Fall off bottom
        if self.ball_y >= self.height-1:
            self.lives -= 1
            self.ball_launched = False
            self.ball_x = self.paddle_x + self.paddle_width//2
            self.ball_y = self.height-3
            self.ball_dx = random.choice([-1, 1])
            self.ball_dy = -1
            if self.lives == 0:
                self.game_over = True
        # Check if all bricks destroyed
        if not any(alive for _, _, alive in self.bricks):
            self.level += 1
            self.init_bricks()
            self.ball_launched = False
            self.ball_x = self.paddle_x + self.paddle_width//2
            self.ball_y = self.height-3
            self.ball_dx = random.choice([-1, 1])
            self.ball_dy = -1

    def move_paddle(self, dx):
        if self.game_over:
            return
        new_x = self.paddle_x + dx
        if 1 <= new_x and new_x + self.paddle_width <= self.width-1:
            self.paddle_x = new_x
            if not self.ball_launched:
                self.ball_x = self.paddle_x + self.paddle_width//2

    def launch_ball(self):
        if self.game_over:
            return
        if not self.ball_launched:
            self.ball_launched = True

    def restart(self):
        self.__init__(self.stdscr)

    def render(self):
        self.stdscr.clear()
        self.draw_border()
        self.draw_paddle()
        self.draw_ball()
        self.draw_bricks()
        self.draw_info()
        self.stdscr.refresh()

    def run(self):
        while self.running:
            key = self.stdscr.getch()
            if key == ord('q') or key == ord('Q'):
                break
            if key == ord('r') or key == ord('R'):
                self.restart()
            if key == ord(' '):
                self.launch_ball()
            if key == curses.KEY_LEFT or key == ord('a') or key == ord('A'):
                self.move_paddle(-2)
            elif key == curses.KEY_RIGHT or key == ord('d') or key == ord('D'):
                self.move_paddle(2)
            self.update_ball()
            self.render()
            curses.napms(50)

def main(stdscr):
    game = Arkanoid(stdscr)
    game.run()

if __name__ == "__main__":
    curses.wrapper(main)
